Refactor helper methods in `DataStreamTestHelper`

nielsbauman opened the pull request:
For most of the usages of these methods, it made more sense to return a `ProjectMetadata` instead of a `ClusterState`.
We also don't need to specify a specific project ID; generating a random one inside the helper method saves some boilerplate code.

nielsbauman committed:
Remove usages of `Metadata#getProject()` in index settings provider tests

Both these test classes required minimal refactoring to get rid of the
deprecated `getProject()` method.

PeteGillinElastic commented on modules/data-streams/src/test/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProviderTests.java:
```diff
@@ -317,18 +318,20 @@ public void testGetAdditionalIndexSettingsDataStreamAlreadyCreated() throws Exce
-        Metadata.Builder mb = Metadata.builder(
+        final var projectId = randomProjectIdOrDefault();
+        ProjectMetadata.Builder projectBuilder = ProjectMetadata.builder(
             DataStreamTestHelper.getClusterStateWithDataStreams(
```
We don't have a `DataStreamTestHelper.getProjectWithDataStreams` that works here, like we had for the `...WithDataStream` (singular) above?

PeteGillinElastic commented on x-pack/plugin/logsdb/src/test/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProviderTests.java:
```diff
@@ -623,16 +622,15 @@ public void testNewIndexHasSyntheticSourceUsageInvalidSettings() throws IOExcept
-        );
-        Metadata metadata = mb.build();
+        final var projectId = randomProjectIdOrDefault();
+        ProjectMetadata project = DataStreamTestHelper.getClusterStateWithDataStreams(
```
Same question as above, would a dedicated helper method to get the `ProjectMetadata` directly be worthwhile?

PeteGillinElastic replied:
I can see a number of other places below that could (it seems) also use such a helper, I'm not going to comment on them all.

PeteGillinElastic reviewed (approved):
LGTM, though I'd consider introducing a new helper method as per my comments, if that works the way I imagine it would.

nielsbauman committed:
Refactor all code in the entire world

nielsbauman committed:
Merge branch 'main' into index-settings-providers

nielsbauman commented:
I have good news and I have bad news. The bad news is that the PR grew quite a bit, but the good news is that the PR looks much better now (or well, the code after merging this PR will look much better). Here's a summary of those changes:
1. I changed the helper methods in `DataStreamTestHelper` to return a `ProjectMetadata` instead of a `ClusterState`.
2. Additionally, those methods now specify a random project ID instead of taking it as a parameter, as (almost) all usages just want a random project ID.
3. Along the way, I updated some tests to use a random project ID instead of the default one, as the production code was already project-aware. I tried to run these tests with a unique project ID to verify that that doesn't break the test, but there's a chance I forgot to double-check some tests. Worst case, a test failure gets opened that tells us the code is not project aware yet.
4. These changes allowed me to remove some redundant cluster state conversion in some tests.
5. Finally, I was able to remove a bunch more `Metadata#getProject()` usages in tests :)

Splitting this PR up into multiple PRs would have been difficult. Fortunately, the changes are pretty independent of one another (except for the `DataStreamTestHelper` of course). Let me know what you think!

PeteGillinElastic commented on test/framework/src/main/java/org/elasticsearch/cluster/metadata/DataStreamTestHelper.java:
```diff
@@ -441,63 +442,35 @@ public static DataStreamGlobalRetention randomGlobalRetention() {
 
-    public static ClusterState getClusterStateWithDataStreams(
-        ProjectId projectId,
+    @FixForMultiProject() // Remove this intermediate method when ProactiveStorageDeciderServiceTests no longer needs the default project ID
```
I'm slightly confused by this comment. This overload isn't the one used in the `@FixForMultiProject`-annotated line in `ProactiveStorageDeciderServiceTests`. That's using an overload which takes the project ID. So I would have expected that the change when that test no longer requires the default project would be to remove (or perhaps make private) the overload which that test currently uses and won't any more?

PeteGillinElastic reviewed (approved):
Hi, sorry for dropping the ball on this review. You should have nudged me earlier!

Anyway, this LGTM, apart from one small question.

nielsbauman committed:
Merge branch 'main' into index-settings-providers

nielsbauman committed:
Move annotation

nielsbauman committed:
Remove unused method
